File: workers/expense-bot/test_apple_pay.js
/**
 * Test Apple Pay integration for Cloudflare Workers bot
 */

import { parseApplePayMessage } from './src/services.js';

/**
 * Test cases for Apple Pay message parsing
 */
const testCases = [
    {
        name: "Valid Apple Pay message with Starbucks",
        input: "Spent $15 at Starbucks on 2026-03-26",
        expected: {
            amount: 15.0,
            merchant: "Starbucks",
            date: "2026-03-26"
        }
    },
    {
        name: "Valid Apple Pay message with McDonald's",
        input: "Spent $5.00 at McDonald's on 2026-03-25",
        expected: {
            amount: 5.0,
            merchant: "McDonald's",
            date: "2026-03-25"
        }
    },
    {
        name: "Valid Apple Pay message with Amazon",
        input: "Spent $100.00 at Amazon on 2026-03-20",
        expected: {
            amount: 100.0,
            merchant: "Amazon",
            date: "2026-03-20"
        }
    },
    {
        name: "Valid Apple Pay message with Vending Machine",
        input: "Spent $0.50 at Vending Machine on 2026-03-26",
        expected: {
            amount: 0.5,
            merchant: "Vending Machine",
            date: "2026-03-26"
        }
    },
    {
        name: "Valid Apple Pay message with Target",
        input: "Spent $25.75 at Target on 2025-12-15",
        expected: {
            amount: 25.75,
            merchant: "Target",
            date: "2025-12-15"
        }
    },
    {
        name: "Valid Apple Pay message with Best Buy",
        input: "Spent $120.00 at Best Buy on 2027-01-01",
        expected: {
            amount: 120.0,
            merchant: "Best Buy",
            date: "2027-01-01"
        }
    },
    {
        name: "Invalid negative amount",
        input: "Spent $-10.00 at Store on 2026-03-26",
        expected: {}
    },
    {
        name: "Invalid non-numeric amount",
        input: "Spent $abc at Store on 2026-03-26",
        expected: {}
    },
    {
        name: "Invalid message format",
        input: "Invalid message format",
        expected: {}
    },
    {
        name: "Missing date",
        input: "Spent $10.00 at Store",
        expected: {}
    },
    {
        name: "Missing date part",
        input: "Spent $10.00 at Store on",
        expected: {}
    },
    {
        name: "Invalid date format (wrong separator)",
        input: "Spent $10.00 at Store on 2026/03/26",
        expected: {}
    },
    {
        name: "Invalid date format (no separators)",
        input: "Spent $10.00 at Store on 20260326",
        expected: {}
    },
    {
        name: "Empty message",
        input: "",
        expected: {}
    },
    {
        name: "Invalid day",
        input: "Spent $15 at Starbucks on 2026-03-32",
        expected: {}
    },
    {
        name: "Invalid month",
        input: "Spent $15 at Starbucks on 2026-13-26",
        expected: {}
    },
    {
        name: "Invalid year (too old)",
        input: "Spent $15 at Starbucks on 1970-02-26",
        expected: {}
    },
    {
        name: "Invalid year (too far future)",
        input: "Spent $15 at Starbucks on 2100-02-26",
        expected: {}
    }
];

/**
 * Run all tests
 */
function runTests() {
    console.log('🧪 Testing Apple Pay Message Parsing');
    console.log('='.repeat(50));
    
    let passed = 0;
    let failed = 0;
    
    testCases.forEach((testCase, index) => {
        const result = parseApplePayMessage(testCase.input);
        const isMatch = JSON.stringify(result) === JSON.stringify(testCase.expected);
        
        if (isMatch) {
            console.log(`✅ Test ${index + 1}: ${testCase.name}`);
            console.log(`   Parsed correctly: ${JSON.stringify(result)}`);
            passed++;
        } else {
            console.log(`❌ Test ${index + 1}: ${testCase.name}`);
            console.log(`   Expected: ${JSON.stringify(testCase.expected)}`);
            console.log(`   Got: ${JSON.stringify(result)}`);
            failed++;
        }
        console.log('');
    });
    
    console.log('='.repeat(50));
    console.log(`📊 Test Results: ${passed} passed, ${failed} failed`);
    
    if (failed === 0) {
        console.log('🎉 All tests passed! Apple Pay integration is ready.');
    } else {
        console.log('❌ Some tests failed. Please check the implementation.');
    }
    
    return failed === 0;
}

// Run tests if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
    runTests();
}

export { runTests };

File: test_apple_pay.py
#!/usr/bin/env python3
"""Test script for Apple Pay message parsing."""

import sys
import os
from pathlib import Path

# Add the expenses-tracker directory to the path
sys.path.insert(0, str(Path(__file__).parent / "expenses-tracker"))

from main import parse_apple_pay_message

def test_apple_pay_parsing():
    """Test the Apple Pay message parsing function."""
    
    print("🧪 Testing Apple Pay Message Parsing")
    print("=" * 50)
    
    # Test cases
    test_cases = [
        # Valid cases
        ("Spent $15 at Starbucks on 26 Mar 2026 at 10:28 PM", True, {
            'amount': 15.0,
            'merchant': 'Starbucks',
            'date': '2026-03-26'
        }),
        ("Spent $5.00 at McDonald's on 25 Mar 2026 at 1:30 PM", True, {
            'amount': 5.00,
            'merchant': "McDonald's",
            'date': '2026-03-25'
        }),
        ("Spent $100.00 at Amazon on 20 Mar 2026 at 11:45 AM", True, {
            'amount': 100.00,
            'merchant': 'Amazon',
            'date': '2026-03-20'
        }),
        ("Spent $0.50 at Vending Machine on 26 Mar 2026 at 9:15 AM", True, {
            'amount': 0.50,
            'merchant': 'Vending Machine',
            'date': '2026-03-26'
        }),
        ("Spent $25.75 at Target on 15 Dec 2025 at 6:20 PM", True, {
            'amount': 25.75,
            'merchant': 'Target',
            'date': '2025-12-15'
        }),
        ("Spent $120.00 at Best Buy on 1 Jan 2027 at 2:00 PM", True, {
            'amount': 120.00,
            'merchant': 'Best Buy',
            'date': '2027-01-01'
        }),
        
        # Invalid cases
        ("Spent $-10.00 at Store on 26 Mar 2026 at 10:28 PM", False, {}),
        ("Spent $abc at Store on 26 Mar 2026 at 10:28 PM", False, {}),
        ("Invalid message format", False, {}),
        ("Spent $10.00 at Store", False, {}),
        ("Spent $10.00 at Store on", False, {}),
        ("Spent $10.00 at Store on 26 March 2026 at 10:28 PM", False, {}),
        ("Spent $10.00 at Store on 26 MAR 2026 at 10:28 PM", False, {}),
        ("", False, {}),
        ("Spent $15 at Starbucks on 32 Mar 2026 at 10:28 PM", False, {}),
        ("Spent $15 at Starbucks on 26 Feb 1970 at 10:28 PM", False, {}),
        ("Spent $15 at Starbucks on 26 Feb 2100 at 10:28 PM", False, {}),
    ]
    
    passed = 0
    failed = 0
    
    for i, (message, should_pass, expected) in enumerate(test_cases, 1):
        print(f"\nTest {i}: {message}")
        
        result = parse_apple_pay_message(message)
        
        if should_pass:
            if result == expected:
                print(f"✅ PASS - Parsed correctly: {result}")
                passed += 1
            else:
                print(f"❌ FAIL - Expected: {expected}, Got: {result}")
                failed += 1
        else:
            if result == {}:
                print(f"✅ PASS - Correctly rejected invalid message")
                passed += 1
            else:
                print(f"❌ FAIL - Should have been rejected, but got: {result}")
                failed += 1
    
    print(f"\n" + "=" * 50)
    print(f"📊 Test Results: {passed} passed, {failed} failed")
    
    if failed == 0:
        print("🎉 All tests passed! Apple Pay integration is ready.")
        return True
    else:
        print("⚠️  Some tests failed. Please check the implementation.")
        return False

if __name__ == "__main__":
    success = test_apple_pay_parsing()
    sys.exit(0 if success else 1)